CC-39249: Document project-defined canonical nested objects (#3777)
spryker/spryker-docs: docs/dg/dev/architecture/api-platform/resource-schemas.md

not a shared class — is how identically-named objects accumulate fields across modules while each
resource keeps its own independent request/response shape.

## Project-defined canonical nested objects

[Typed nested objects](#typed-nested-objects) generate one value-object class **per resource
property**: a `billingAddress` on the checkout resource and a `shippingAddress` on the order
resource each get their own independent class, even when both describe the same real-world shape.
That keeps each resource self-contained, but it also means the same address shape is authored and

> maintained in several places.

Reviewer: Did we produce canonical names for all migrated endpoints?
I believe that to make them BC, they must contain canonical names.

Author: The endpoints don't need extra canonical names; they are merged from module A and module B into one class.

We only add objectName to the nested object definitions.

A **canonical nested object** lets a project define that shared shape **once** and have it flow
into every resource property that opts in. All the opting-in properties then collapse onto a single
generated class — `Generated\Api\{ApiType}\{Object}` (for example, `Generated\Api\Storefront\Address`) —
instead of a per-resource companion class.

> This is a pure project opt-in. With no canonical object files present, generation is byte-for-byte

Reviewer: ah, so we didn't use canonical names for migrated endpoints.

The `<dashed-name>.<kind>.yml` naming pattern is the same for both file types — only the kind word differs. `address.object.yml` is the canonical-object analog of `checkout.resource.yml`, and `address.object.validation.yml` is the analog of `checkout.validation.yml`. The `object` versus `resource` word identifies the artifact kind, not a different naming scheme.

Full path patterns:

```text
resources/api/<apiType>/objects/<dashed-name>.object.yml
resources/api/<apiType>/objects/<dashed-name>.object.validation.yml   # optional, see Validation
```

For example, on the storefront API:

```text
src/Pyz/resources/api/storefront/objects/address.object.yml
src/Pyz/resources/api/storefront/objects/address.object.validation.yml
```

The file name uses a dashed (kebab-case) object name, while `object.name` **inside** the file is
CamelCase. The CamelCase `object.name` is the contract: it must exactly match the `objectName:`
join tag declared on the resource properties that want this shape (see [The `objectName` join
tag](#the-objectname-join-tag)).

### Central directory

> A project may keep canonical object files in one central location instead of (or in addition to) the per-module `objects/` directories. Both locations are scanned simultaneously.

Reviewer: Why may project developer want this?

Author: You know projects better than I do. I would find it easier to use when I have those canonical definitions in one place instead of spread across project modules.

Configure the central directory via the Symfony bundle config node `spryker_api_platform.canonical_object_search_directories`, keyed by API type. Relative paths resolve against the project root; `%kernel.project_dir%` is also supported:

```yaml
# config/packages/spryker_api_platform.yaml
spryker_api_platform:
    canonical_object_search_directories:
        storefront:
            - '%kernel.project_dir%/config/api/objects/storefront'
```

The same `*.object.yml` / `*.object.validation.yml` naming rules apply. Files in a central directory are always treated as the **project** layer, so they participate in the standard `project > feature > core` merge precedence.

Defining the same `objectName` more than once within the same layer — for example, one module file and one central-directory file both at project layer — is a fail-loud error: generation aborts with an `ApiSchemaGenerationException` naming both source files. The same name across different layers is fine — that is the normal override.

### File format

The file contains a single top-level `object:` key:

```yaml
# address.object.yml
object:

>     name: Address                                   # CamelCase; matches `objectName: Address` on resource properties

Reviewer: ```suggestion
    name: Address                                   # CamelCase; produces `objectName: Address` on resource properties
```
I'm trying to understand - what comes first.

Author: This part is really about matching. The canonical definitions name, e.g. Address matches what the resource schemas have as objectName.

You define a resource with a nested resource and you give it an objectName. When projects add the canonical definition and they use a name matching what is defined in the resource schema this canonical object will then be used.

    properties:
        salutation: { type: string, description: 'Address salutation.', example: 'Mr' }
        firstName:  { type: string, description: 'First name.', example: 'Jane' }
        lastName:   { type: string, description: 'Last name.', example: 'Doe' }
        address1:   { type: string, description: 'Street name.', example: 'Julie-Wolfthorn-Straße' }
        zipCode:    { type: string, description: 'ZIP / postal code.', example: '10115' }
        city:       { type: string, description: 'City.', example: 'Berlin' }
```

| Key | Type | Required | Description |
|-----|------|----------|-------------|

> | `object.name` | string | Yes | CamelCase object name. Must match the `objectName:` join tag on every resource property that references this object. |

Reviewer: ```suggestion
| `object.name` | string | Yes | CamelCase object name. Matched against `objectName:` join tag on every resource property that references this object. |
```
